### PR TITLE
Iox #542 sigaction abstraction

### LINKED ISSUE #542: Fix resource leak caused by move ctor/assignment
## Required information

All versions

**Observed result or behaviour:**
The classes:
 - [x] `unix_domain_socket` (fixed by https://github.com/eclipse-iceoryx/iceoryx/pull/550)
 - [x] `shared_memory`
 - [x] `memory_map`
 - [x] `GenericRAII`

are not cleaning up the resources while performing move. This could lead to a resource leak when for instance the following pseudo code is executed.

```cpp
SharedMemory mySharedMemory; // create a shared memory file A
SharedMemory anotherSharedMemory; // create a shared memory file B

mySharedMemory = std::move(anotherSharedMemory); // mySharedMemory takes ownership of B without releasing A

// at this point the shared memory file A is leaked

```

**Classes with correct move resource handling:**
- `Semaphore`
- `MessageQueue`
- `optional`
- `variant`

**Handle undefined state:**
- [x] `optional`
- [x] `expected`

**creation pattern:**
- [x] `shared_memory`
- [x] `shared_memory_object`
- [x] `memory_map`

**Fix incorrect move CTor/assigned in all classes derived from `DesignPattern::Creation`**
According to: https://isocpp.org/wiki/faq/assignment-operators every move ctor/assignment operator has to be implemented like
```cpp
Derived& Derived::operator= (Derived&& d)
{
  // self-assignment is not allowed in move assignment
  Base::operator= (std::move(d));
  // Do the rest of your assignment operator here...
  return *this;
}
```
- [x] `shared_memory`
- [x] `shared_memory_object`
- [x] `memory_map`
- [x] `semaphore`
- [x] `message_queue`
- [x] `unix_domain_socket`
- [x] Write a detailed documentation how to use the creation pattern correctly - especially the assignment operator part!

**Fix sighandler**
- [x] Write mini `sigaction` abstraction and put it into utils to remove the 5 lines of code.
- [x] Use `posix::SignalHandler` everywhere

### Details
* Consider [this](https://github.com/eclipse-iceoryx/iceoryx/pull/550#discussion_r573629792) discussion

### PR DESCRIPTION
**Pull request by elfenpiff**
## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [x] All checkboxes in the PR checklist are checked or crossed out
1. [x] Merge

## References

- Partially closes #542

### CODE REVIEW

File: iceoryx_utils/source/posix_wrapper/signal_handler.cpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                  << std::endl;
+        return SignalGuard();
+    }
+
+    action.sa_handler = callback;
+    action.sa_flags = 0;
+
+    struct sigaction previousAction;
+
+    if (cxx::makeSmartC(sigaction,
```
Comment (dkroenke) [inline iceoryx_utils/source/posix_wrapper/signal_handler.cpp:93]: It should be considered that if you register a sigaction multiple times for the same signal () that only the last registered handler is used.

Comment (elfenpiff) [reply iceoryx_utils/source/posix_wrapper/signal_handler.cpp:93]: This is why I wrote the @attention NEVER USE THIS CLASS AS A MEMBER in the doxygen docu. I also described the use case there as well and with this I think this is covered.

File: iceoryx_utils/include/iceoryx_utils/posix_wrapper/signal_handler.hpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+/// @attention NEVER USE THIS CLASS AS A MEMBER VARIABLE! A class which should be used
+///            only in method/function scopes.
+/// @brief The SignalGuard is a class returned by registerSignalHandler. When it goes
+///         out of scope it restores the previous signal action. Typical use case:
+///         One would like to override the signal action in main() or some C posix
+///         makes it necessary to override the standard signal action before and
+///         after the call.
+/// @code
+///    {
+///      auto sigHandler = registerSignalHandler(Signal::BUS, printErrorMessage);
```
Comment (elBoberido) [inline iceoryx_utils/include/iceoryx_utils/posix_wrapper/signal_handler.hpp:47]: this might be confusing since you get the `SignalGuard` not the signal handler

File: iceoryx_utils/include/iceoryx_utils/posix_wrapper/signal_handler.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+///      my_c_call_which_can_cause_SIGBUS();
+///    }
+///    // here we are out of scope and the signal action for Signal::BUS is restored
+/// @endcode
+class SignalGuard
+{
+  public:
+    SignalGuard() noexcept = default;
+    SignalGuard(const Signal signal, const struct sigaction& previousAction) noexcept;
```
Comment (elBoberido) [inline iceoryx_utils/include/iceoryx_utils/posix_wrapper/signal_handler.hpp:56]: Please add documentation.

Should this be public? Maybe it's less error prone if it can only be used by `registerSignalHandler´

File: iceoryx_utils/source/posix_wrapper/signal_handler.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    {
+        restorePreviousAction();
+
+        m_signal = std::move(rhs.m_signal);
+        m_previousAction = std::move(rhs.m_previousAction);
+        m_doRestorePreviousAction = std::move(rhs.m_doRestorePreviousAction);
+
+        rhs.m_doRestorePreviousAction = false;
+    }
+    return *this;
```
Comment (elBoberido) [inline iceoryx_utils/source/posix_wrapper/signal_handler.cpp:53]: It might be surprising that the move restores the previous action. It is technically correct, but still surprising. I would delete the move assignment and allow only the move ctor.

**Review by elBoberido — COMMENTED** [review]
I did not yet review the tests.
Regarding the code. Let's assume I want to register `SIGINT` and `SIGTERM` in a function. An because I don't like two variables I create a `cxx::vector` and do an `emplace`. In which order are they destroyed and which sigaction will be restored at last?

File: iceoryx_utils/source/posix_wrapper/signal_handler.cpp (hunk 1 above)
Comment (dkroenke) [reply iceoryx_utils/source/posix_wrapper/signal_handler.cpp:93]: I'm referring to the case that `registerSignalHandler` overwrite the previous set `sa_handler` silently when you call it multiple times. So either we state that clearly in the documentation or we need to prevent that.

**Comment by elfenpiff** [comment]
> I did not yet review the tests.
> Regarding the code. Let's assume I want to register `SIGINT` and `SIGTERM` in a function. An because I don't like two variables I create a `cxx::vector` and do an `emplace`. In which order are they destroyed and which sigaction will be restored at last?

Great question. Since they are two different signals both will be restored to their original state!

**Review by dkroenke — DISMISSED** [review]

**Comment by elBoberido** [comment]
> > I did not yet review the tests.
> > Regarding the code. Let's assume I want to register `SIGINT` and `SIGTERM` in a function. An because I don't like two variables I create a `cxx::vector` and do an `emplace`. In which order are they destroyed and which sigaction will be restored at last?
> 
> Great question. Since they are two different signals both will be restored to their original state!

Of course :D
What I was curious about was in which order the `cxx::vector` destroys it's elements. Let's assume you register `SIGINT` twice and store it in a `cxx::vector`. Are they destroyed in reverse order of creation? It's a misuse in this case, but what about other classes stored in a `cxx::vector`? If it is not destroyed in reverse order, is it okay?

File: iceoryx_utils/test/moduletests/test_posix_signal_handler.cpp [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_utils/posix_wrapper/signal_handler.hpp"
+#include "test.hpp"
+
+using namespace ::testing;
+using namespace iox::posix;
+
+int signalOfCallback1 = 0;
+int signalOfCallback2 = 0;
```
Comment (elBoberido) [inline iceoryx_utils/test/moduletests/test_posix_signal_handler.cpp:24]: Shouldn't this be atomics? I'm not quite sure, but since it is set from a different context it might be necessary.

Comment (MatthiasKillat) [reply iceoryx_utils/test/moduletests/test_posix_signal_handler.cpp:24]: I am also not quite sure. I think the kernel causes the signalHandler to be executed in the thread which receives the signal, so an atomic would not be necessary here. A quick search confirms this, but it is not a reliable source https://stackoverflow.com/questions/6949025/how-are-asynchronous-signal-handlers-executed-on-linux.

Would not hurt to make those atomic I guess, then we would be sure the updates will be visible here as well (after some cache syncs at least....).

**Comment by MatthiasKillat** [comment]
@elBoberido  According to https://stackoverflow.com/questions/6169125/order-of-destruction-of-elements-of-an-stdvector which refers to the standard the elements are destroyed in reverse order.

I would consider this a misuse by the user, she needs to consider destruction order of the container when using such scope guards. (It might get much less predictable with structures using trees for internal representation, e.g. maps, the user needs to consider this)

TLDR: not the responsibility of this class.

**Review by dkroenke — APPROVED** [review]

**Review by MatthiasKillat — APPROVED** [review]
